File: composeApp/src/commonMain/kotlin/com/grateful/deadly/feature/recording/components/RecordingSelectionSheet.kt
package com.grateful.deadly.feature.recording.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Button
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.SheetState
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.grateful.deadly.core.design.icons.AppIcon
import com.grateful.deadly.core.design.icons.Render
import com.grateful.deadly.domain.models.RecordingSelectionAction
import com.grateful.deadly.domain.models.RecordingSelectionState

/**
 * RecordingSelectionSheet - Modal bottom sheet for recording selection
 *
 * Follows V2's design patterns:
 * - Material3 ModalBottomSheet with skipPartiallyExpanded
 * - Header with close button and show info
 * - Scrollable list of RecordingOptionCard components
 * - Action buttons for "Set as Default" and "Reset to Recommended"
 * - Loading and error states
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RecordingSelectionSheet(
    state: RecordingSelectionState,
    onAction: (RecordingSelectionAction) -> Unit,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
    sheetState: SheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
) {
    if (state.isVisible) {
        ModalBottomSheet(
            onDismissRequest = onDismiss,
            sheetState = sheetState,
            modifier = modifier
        ) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(16.dp)
            ) {
                // Header with close button and show info
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(
                            text = "Choose Recording",
                            style = MaterialTheme.typography.headlineSmall,
                            fontWeight = FontWeight.Bold
                        )
                        if (state.showTitle.isNotBlank()) {
                            Text(
                                text = state.showTitle,
                                style = MaterialTheme.typography.bodyMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                        if (state.showDate.isNotBlank()) {
                            Text(
                                text = state.showDate,
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant
                            )
                        }
                    }

                    IconButton(onClick = onDismiss) {
                        AppIcon.Close.Render(
                            size = 24.dp,
                            tint = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }

                Spacer(modifier = Modifier.height(16.dp))

                // Content based on state
                when {
                    state.isLoading -> {
                        LoadingContent()
                    }

                    state.errorMessage != null -> {
                        ErrorContent(
                            message = state.errorMessage,
                            onRetry = { /* TODO: Add retry action */ }
                        )
                    }

                    else -> {
                        RecordingListContent(
                            state = state,
                            onAction = onAction
                        )
                    }
                }

                Spacer(modifier = Modifier.height(16.dp))

                // Action buttons
                ActionButtons(
                    state = state,
                    onAction = onAction
                )

                // Bottom padding for safe area
                Spacer(modifier = Modifier.height(16.dp))
            }
        }
    }
}

@Composable
private fun LoadingContent() {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(200.dp),
        contentAlignment = Alignment.Center
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            CircularProgressIndicator()
            Text(
                text = "Loading recordings...",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant
            )
        }
    }
}

@Composable
private fun ErrorContent(
    message: String,
    onRetry: () -> Unit
) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .height(200.dp),
        contentAlignment = Alignment.Center
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Text(
                text = "Error loading recordings",
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.error
            )
            Text(
                text = message,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center
            )
            OutlinedButton(onClick = onRetry) {
                Text("Retry")
            }
        }
    }
}

@Composable
private fun RecordingListContent(
    state: RecordingSelectionState,
    onAction: (RecordingSelectionAction) -> Unit
) {
    LazyColumn(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        items(
            items = state.allRecordings,
            key = { it.identifier }
        ) { recording ->
            RecordingOptionCard(
                recordingOption = recording,
                onClick = {
                    onAction(RecordingSelectionAction.SelectRecording(recording.identifier))
                }
            )
        }

        // Empty state if no recordings
        if (state.allRecordings.isEmpty()) {
            item {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .height(100.dp),
                    contentAlignment = Alignment.Center
                ) {
                    Text(
                        text = "No recordings available",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        }
    }
}

@Composable
private fun ActionButtons(
    state: RecordingSelectionState,
    onAction: (RecordingSelectionAction) -> Unit
) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(12.dp)
    ) {
        // Reset to Recommended button
        if (state.shouldShowResetToRecommended) {
            OutlinedButton(
                onClick = { onAction(RecordingSelectionAction.ResetToRecommended) },
                modifier = Modifier.weight(1f)
            ) {
                Text("Reset to Recommended")
            }
        }

        // Set as Default button
        if (state.shouldShowSetAsDefault) {
            val selectedRecording = state.selectedRecording
            if (selectedRecording != null) {
                Button(
                    onClick = {
                        onAction(RecordingSelectionAction.SetAsDefault(selectedRecording.identifier))
                    },
                    modifier = Modifier.weight(1f)
                ) {
                    Text("Set as Default")
                }
            }
        }

        // If no action buttons are shown, add spacer
        if (!state.shouldShowResetToRecommended && !state.shouldShowSetAsDefault) {
            Spacer(modifier = Modifier.width(1.dp))
        }
    }
}

/**
 * Preview functions for interactive testing
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RecordingSelectionSheet_Preview_Loading() {
    MaterialTheme {
        RecordingSelectionSheet(
            state = RecordingSelectionState(
                isVisible = true,
                showTitle = "Grateful Dead Live at Cornell University",
                showDate = "May 8, 1977",
                isLoading = true
            ),
            onAction = {},
            onDismiss = {}
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RecordingSelectionSheet_Preview_WithRecordings() {
    MaterialTheme {
        RecordingSelectionSheet(
            state = RecordingSelectionState(
                isVisible = true,
                showTitle = "Grateful Dead Live at Cornell University",
                showDate = "May 8, 1977",
                currentRecording = com.grateful.deadly.domain.models.RecordingOptionViewModel(
                    identifier = "gd1977-05-08.sbd.hicks.4982.sbeok.shnf",
                    sourceType = "SBD",
                    taperInfo = "Taper: Betty Cantor",
                    technicalDetails = "Soundboard, DAT → CD → FLAC",
                    rating = 4.8f,
                    reviewCount = 23,
                    rawSource = "Betty Soundboard",
                    rawLineage = "SBD > DAT > CD > FLAC",
                    isSelected = true,
                    isCurrent = true,
                    isRecommended = true,
                    matchReason = "Recommended"
                ),
                alternativeRecordings = listOf(
                    com.grateful.deadly.domain.models.RecordingOptionViewModel(
                        identifier = "gd1977-05-08.aud.miller.23456.flac16",
                        sourceType = "AUD",
                        taperInfo = "Taper: Mike Miller",
                        technicalDetails = "Audience, Nak 700 → Cassette → DAT",
                        rating = 4.2f,
                        reviewCount = 12,
                        rawSource = "Audience",
                        rawLineage = "AUD > Cassette > DAT > FLAC",
                        isSelected = false,
                        isCurrent = false,
                        isRecommended = false,
                        matchReason = "Popular Choice"
                    ),
                    com.grateful.deadly.domain.models.RecordingOptionViewModel(
                        identifier = "gd1977-05-08.mtx.jones.11111",
                        sourceType = "Matrix",
                        taperInfo = null,
                        technicalDetails = "Matrix blend of SBD + AUD",
                        rating = null,
                        reviewCount = null,
                        rawSource = null,
                        rawLineage = null,
                        isSelected = false,
                        isCurrent = false,
                        isRecommended = false,
                        matchReason = null
                    )
                ),
                hasRecommended = true
            ),
            onAction = {},
            onDismiss = {}
        )
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RecordingSelectionSheet_Preview_Error() {
    MaterialTheme {
        RecordingSelectionSheet(
            state = RecordingSelectionState(
                isVisible = true,
                showTitle = "Grateful Dead Live at Cornell University",
                showDate = "May 8, 1977",
                errorMessage = "Failed to load recordings from Archive.org"
            ),
            onAction = {},
            onDismiss = {}
        )
    }
}

File: iosApp/iosApp/Views/ShowDetail/RecordingSelectionSheet.swift
import SwiftUI
import ComposeApp

/**
 * RecordingSelectionSheet - Modal bottom sheet for recording selection
 *
 * Matches Compose RecordingSelectionSheet.kt exactly:
 * - Header with title, show info, and close button
 * - Scrollable list of RecordingOptionCardView items
 * - Loading and error states
 * - Dynamic action buttons:
 *   - "Reset to Recommended" (outlined) - only when not using recommended
 *   - "Set Recording for Show" (filled) - only when different recording selected
 */

private let DeadRed = Color(red: 0xDC/255.0, green: 0x14/255.0, blue: 0x3C/255.0)

struct RecordingSelectionSheet: View {
    @ObservedObject var viewModel: RecordingSelectionViewModelWrapper
    let onDismiss: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            // Header with close button and show info
            HStack(alignment: .top) {
                VStack(alignment: .leading, spacing: 4) {
                    Text("Choose Recording")
                        .font(.title2)
                        .fontWeight(.bold)

                    if !viewModel.showTitle.isEmpty {
                        Text(viewModel.showTitle)
                            .font(.body)
                            .foregroundColor(.secondary)
                    }

                    if !viewModel.showDate.isEmpty {
                        Text(viewModel.showDate)
                            .font(.caption)
                            .foregroundColor(.secondary)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)

                Button(action: onDismiss) {
                    Image(systemName: "xmark")
                        .font(.system(size: 16))
                        .foregroundColor(.secondary)
                        .frame(width: 32, height: 32)
                }
            }
            .padding(.horizontal, 16)
            .padding(.top, 16)

            // Content based on state
            if viewModel.isLoading {
                loadingContent
            } else if let errorMessage = viewModel.errorMessage {
                errorContent(message: errorMessage)
            } else {
                recordingListContent
            }

            // Action buttons - always visible at bottom
            actionButtons
                .padding(.horizontal, 16)
                .padding(.bottom, 32) // Extra bottom padding for safe area
        }
        .presentationDetents([.medium, .large])
        .presentationDragIndicator(.visible)
    }

    // MARK: - Loading State

    private var loadingContent: some View {
        VStack(spacing: 16) {
            ProgressView()
            Text("Loading recordings...")
                .font(.body)
                .foregroundColor(.secondary)
        }
        .frame(maxWidth: .infinity)
        .frame(height: 200)
    }

    // MARK: - Error State

    private func errorContent(message: String) -> some View {
        VStack(spacing: 16) {
            Text("Error loading recordings")
                .font(.headline)
                .foregroundColor(.red)

            Text(message)
                .font(.body)
                .foregroundColor(.secondary)
                .multilineTextAlignment(.center)

            Button("Retry") {
                // TODO: Add retry action
            }
            .buttonStyle(.bordered)
        }
        .frame(maxWidth: .infinity)
        .frame(height: 200)
        .padding(.horizontal, 16)
    }

    // MARK: - Recording List Content

    private var recordingListContent: some View {
        ScrollView {
            VStack(spacing: 8) {
                // Current recording first (if exists)
                if let currentRecording = viewModel.currentRecording {
                    RecordingOptionCardView(
                        recording: currentRecording,
                        onClick: {
                            viewModel.selectRecording(currentRecording.identifier)
                        }
                    )
                }

                // Alternative recordings
                ForEach(viewModel.alternativeRecordings, id: \.identifier) { recording in
                    RecordingOptionCardView(
                        recording: recording,
                        onClick: {
                            viewModel.selectRecording(recording.identifier)
                        }
                    )
                }

                // Empty state
                if viewModel.allRecordings.isEmpty {
                    Text("No recordings available")
                        .font(.body)
                        .foregroundColor(.secondary)
                        .frame(maxWidth: .infinity)
                        .frame(height: 100)
                }
            }
            .padding(.horizontal, 16)
        }
        .frame(maxHeight: 400) // Max height constraint like Compose
    }

    // MARK: - Action Buttons

    private var actionButtons: some View {
        VStack(spacing: 12) {
            // Reset to Recommended button (only when not using recommended)
            if viewModel.shouldShowResetToRecommended {
                Button(action: {
                    viewModel.resetToRecommended()
                    onDismiss()
                }) {
                    HStack(spacing: 8) {
                        Image(systemName: "star")
                            .font(.system(size: 16))
                        Text("Reset to Recommended")
                            .font(.body)
                    }
                    .frame(maxWidth: .infinity)
                    .frame(height: 48)
                }
                .buttonStyle(.bordered)
            }

            // Set as Default button (only when different recording selected)
            if viewModel.shouldShowSetAsDefault,
               let selectedRecording = viewModel.selectedRecording {
                Button(action: {
                    viewModel.setAsDefault(selectedRecording.identifier)
                    onDismiss()
                }) {
                    HStack(spacing: 8) {
                        Image(systemName: "star.fill")
                            .font(.system(size: 20))
                        Text("Set Recording for Show")
                            .font(.body)
                    }
                    .frame(maxWidth: .infinity)
                    .frame(height: 48)
                }
                .buttonStyle(.borderedProminent)
                .tint(DeadRed)
            }
        }
    }
}

// MARK: - Preview

#Preview {
    // Preview disabled - requires Koin initialization
    Text("RecordingSelectionSheet Preview")
}
